_posts/2024-12-02-safety-of-methods-for-numeric-primitive-types.md

```diff
@@ -0,0 +1,195 @@
+
+3. **Using Macros for Code Generation**: To efficiently handle the large number of methods and types that required verification, we utilized Rust macros. These macros automated the generation of proof harnesses, reducing redundancy and improving maintainability.
+
+4. **Handling Large Input Spaces**: For methods with large input spaces (e.g., `unchecked_mul` for 64-bit integers), we made the verification process more tractable by:
+   - Introducing assumptions to limit the range of inputs.
+   - Focusing on specific edge cases (e.g., maximum values, zero, and boundary conditions) that are most likely to cause overflow or underflow.
+
+We will walk through each part with examples from the Unsafe Integer Methods part of the challenge.
+
+### 1. Specifying Safety Preconditions
+```rust
+#[requires(!self.overflowing_add(rhs).1)]
+pub const unsafe fn unchecked_add(self, rhs: Self) -> Self {
+    assert_unsafe_precondition!(
+        check_language_ub,
+        concat!(stringify!($SelfT), "::unchecked_add cannot overflow"),
+        (
+            lhs: $SelfT = self,
+            rhs: $SelfT = rhs,
+        ) => !lhs.overflowing_add(rhs).1,
+    );
+
+    // SAFETY: The caller guarantees that the preconditions are met.
+    unsafe {
+        intrinsics::unchecked_add(self, rhs)
+    }
+}
+```
+Here, the ```#[requires]``` attribute is used to specify that `unchecked_add` requires the addition to not overflow. This ensures that the function operates within defined behavior when called in an unsafe context. While we directly leveraged the existing precondition in `unchecked_add`, not every function has an explicit precondition check. Still, we can write preconditions for unsafe functions based on the `SAFETY` section in the official documentation or the `SAFETY` comments in the source code.
+
+The same logic applies to the postconditions (`#[ensure]`) of a function.
+
+### 2. Generating Verification Harnesses
+We created macros to generate verification harnesses for each method and type. For example, to generate harnesses for ```unchecked_add```:
+```rust
+#[cfg(kani)]
+mod verify {
+    use super::*;
+
+    macro_rules! generate_unchecked_math_harness {
+        ($type:ty, $method:ident, $harness_name:ident) => {
+            #[kani::proof_for_contract($type::$method)]
+            pub fn $harness_name() {
+                let num1: $type = kani::any::<$type>();
+                let num2: $type = kani::any::<$type>();
+
+                unsafe {
+                    num1.$method(num2);
+                }
+            }
+        }
+    }
+
+    generate_unchecked_math_harness!(i8, unchecked_add, unchecked_add_i8);
+    generate_unchecked_math_harness!(i16, unchecked_add, unchecked_add_i16);
+    // ... Repeat for other integer types
+}
+```
+The harness contains several parts:
+- Symbolic values: These values (```num1``` and ```num2```) are generated by Kani using ```kani::any```.
```

Comment:
Perhaps add a reference to https://model-checking.github.io/kani/tutorial-nondeterministic-variables.html.